Python: document api graphs (#5425)
github/codeql: docs/codeql/codeql-language-guides/using-api-graphs-in-python.rst

.. _using-api-graphs-in-python:

Using API graphs in Python
==========================

API graphs are a uniform interface for referring to functions, classes, and methods defined in
external libraries.

About this article
------------------

This article describes how to use API graphs to reference classes and functions defined in library
code. This can be used to conveniently refer to external library functions when defining things like
remote flow sources.

Module imports
--------------

The most common entry point into the API graph will be the point where an external module or package is
imported. The API graph node corresponding to the ``re`` library, for instance, can be accessed
using the ``API::moduleImport`` method defined in the ``semmle.python.ApiGraphs`` module, as the
following snippet demonstrates.

.. code-block:: ql

    import python
    import semmle.python.ApiGraphs

    select API::moduleImport("re")

> ➤ `See this in the query console on LGTM.com <https://lgtm.com/query/1876172022264324639/>`__.

Author: I think perhaps the API node corresponding to the imported module does not have a location and therefor does not show up as a result. While `moduleImport` is not meant to be used on its own, I wonder if this behaviour should be considered a bug, since we might be tempted to detect if certain libraries are imported just by seeing if `moduleImport` gives results..

> ➤ `See this in the query console on LGTM.com <https://lgtm.com/query/1876172022264324639/>`__.

Author: Issue created.

.. code-block:: ql

    import python
    import semmle.python.ApiGraphs

    select API::moduleImport("re").getMember("compile").getReturn().getAUse()

➤ `See this in the query console on LGTM.com <https://lgtm.com/query/4346050399960356921/>`__.

Note that this includes all uses of the result of ``re.compile``, including those reachable via
local flow. To get just the *calls* to ``re.compile``, we can use ``getAnImmediateUse`` instead of
``getAUse``. As this is a common occurrence, the method ``getACall`` can be used instead of
``getReturn`` followed by ``getAnImmediateUse``. 

➤ `See this in the query console on LGTM.com <https://lgtm.com/query/8143347716552092926/>`__.

Note that the API graph does not distinguish between class instantiations and function calls. As far
as it's concerned, both are simply places where an API graph node is called.

Subclasses
----------

For many libraries, the main mode of usage is to extend one or more library classes. To track this
in the API graph, we can use the ``getASubclass`` method to get the API graph node corresponding to
all the immediate subclasses of this node. To find *all* subclasses, use ``*`` or ``+`` to apply the

> method repeatedly, as in `getASubclass*`.

Author: Good catch on the quotes.

Note that ``getASubclass`` does not account for any subclassing that takes place in library code
that has not been extracted. Thus, it may be necessary to account for this in the models you write.
For example, the ``flask.views.View`` class has a predefined subclass ``MethodView``, and so to find
all subclasses of ``View``, we must explicitly include the subclasses of ``MethodView`` as well.

.. code-block:: ql

    import python
    import semmle.python.ApiGraphs

    API::Node viewClass() {
      result =
        API::moduleImport("flask").getMember("views").getMember(["View", "MethodView"]).getASubclass*()
    }
    

>     select viewClass()

Author: The suggested one is what would be used in query writing, so let us go with that one now. I do not think there is a great difference in didactic value...

➤ `See this in the query console on LGTM.com <https://lgtm.com/query/288293322319747121/>`__.

Note the use of the set literal ``["View", "MethodView"]`` to match both classes simultaneously.

Built-in functions and classes
------------------------------

> Built-in functions and classes can be accessed using the ``API::builtin`` method, giving the name of
> the built-in as an argument.

Reviewer: ```suggestion
You can access built-in functions and classes using the ``API::builtin`` method, giving the name of
the built-in as an argument.
```

> To find all calls to the built-in ``open`` function, for instance, can be done using the following snippet

Reviewer: ```suggestion
For example, to find all calls to the built-in ``open`` function, you can use the following snippet.
```
